### PR TITLE
feat: create diva subgraph

### PR DESCRIPTION
**Pull request by KedziaPawel**

### CODE REVIEW

File: subgraphs/diva-staking/mappings/ComptrollerLib.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  createDeposit(event.params.buyer, tranches, event);
+
+  let depositor = getDepositor(event.params.buyer);
+
+  if (depositor) {
+    updateDepositor(depositor, tranches, event.params.sharesReceived, event);
+  } else {
+    createDepositor(event.params.buyer, tranches, event.params.sharesReceived, event);
+    updateDepositorCounter(1, event); // increase by 1
+  }
```
Comment (iherger) [inline subgraphs/diva-staking/mappings/ComptrollerLib.ts:36]: I would create a function `ensureDepositor`, like we do in many other places.

File: subgraphs/diva-staking/mappings/ComptrollerLib.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+export function handleSharesRedeemed(event: SharesRedeemed): void {
+  let depositor = getDepositor(event.params.redeemer);
+
+  if (depositor == null) {
+    return;
+  }
```
Comment (iherger) [inline subgraphs/diva-staking/mappings/ComptrollerLib.ts:44]: Create a function `useDepositor`, like we do in many other places. That function should throw if the depositor is not found.

File: subgraphs/diva-staking/mappings/ComptrollerLib.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  // assumption that vault will redeem ETH/stETH/divaETH
+  let redeemAmount = event.params.receivedAssetAmounts[0];
```
Comment (iherger) [inline subgraphs/diva-staking/mappings/ComptrollerLib.ts:47]: We should handle all cases, in particular also the case, where redemptions are in kind and there are multiple assets in the vault (e.g. stETH and divaETH).

File: subgraphs/diva-staking/mappings/ComptrollerLib.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  let tranches = getRedemptionTranches(depositor.trancheAmounts, redeemAmount);
+  let accruedRewards = getAccruedRewards(event.block.timestamp, tranches);
+
+  createRedemption(event.params.redeemer, tranches, accruedRewards, event);
+
+  let updatedDepositor = updateDepositor(depositor, tranches, event.params.sharesAmount.neg(), event);
```
Comment (iherger) [inline subgraphs/diva-staking/mappings/ComptrollerLib.ts:54]: I would do the update inline.

File: subgraphs/diva-staking/mappings/ComptrollerLib.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  if (updatedDepositor.shares.isZero()) {
+    store.remove('Depositor', updatedDepositor.id);
+    updateDepositorCounter(-1, event); // decrease by 1
+  }
```
Comment (iherger) [inline subgraphs/diva-staking/mappings/ComptrollerLib.ts:59]: I would keep the depositor.

File: subgraphs/diva-staking/utils/calcVaultsGav.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  let vaultStEth = ProtocolSdk.bind(stEthVaultAddress);
+  let vaultEth = ProtocolSdk.bind(ethVaultAddress);
+
+  let vaultStEthAccessor = ProtocolSdk.bind(vaultStEth.getAccessor());
+  let vaultEthAccessor = ProtocolSdk.bind(vaultEth.getAccessor());
+
+  let vaultStEthGav = vaultStEthAccessor.calcGav();
+  let vaultEthGav = vaultEthAccessor.calcGav();
+
+  return vaultStEthGav.plus(vaultEthGav);
```
Comment (iherger) [inline subgraphs/diva-staking/utils/calcVaultsGav.ts:15]: Add a comment that we are using 1 ETH = 1 stETH

File: subgraphs/diva-staking/schema.graphql [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  trancheAmounts: [BigInt!]!
+  updatedAt: Int!
+  createdAt: Int!
+  shares: BigInt!
+}
+
+type Redemption @entity {
+  id: ID!
+  redeemer: Bytes!
+  amounts: [BigInt!]! # amounts are stored as negatives
```
Comment (iherger) [inline subgraphs/diva-staking/schema.graphql:31]: I would store redemption amounts as positive.

File: subgraphs/diva-staking/utils/tranches.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+  let scale = 10 ** 6;
+  let scaleBigInt = BigInt.fromU64(scale);
+
+  if (daysStaked <= cooldownDays) {
+    // first half is available after cooldown, and second after staking ends
+    let firstClaimAmount = amount
+      .times(BigInt.fromU64((divPerEthPerDay * daysStaked * scale) as u64).div(BigInt.fromI32(2)))
+      .div(scaleBigInt);
+
+    return new Claim(firstClaimAmount, firstClaimAmount);
```
Comment (iherger) [inline subgraphs/diva-staking/utils/tranches.ts:148]: Shouldn't this be `Claim(firstClaimAmount, 0)`?

File: subgraphs/diva-staking/utils/vaultHandlers.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import { createComptroller } from '../entities/Comptroller';
+
+import { ComptrollerLibDataSource } from '../generated/templates';
+import { Address } from '@graphprotocol/graph-ts';
+
+export function handleAccessorSet(nextAccessor: Address, vault: Address): void {
+  createComptroller(nextAccessor, vault);
+  ComptrollerLibDataSource.create(nextAccessor);
+}
```
Comment (iherger) [inline subgraphs/diva-staking/utils/vaultHandlers.ts:9]: I would remove this file, and just add the two lines in the respective mapping files. Less code, even if duplicated.

**Review by iherger — CHANGES_REQUESTED** [review]
Looks good! I left some comments apart from the migration to use `BigDecimal` that we already talked about.

File: subgraphs/diva-staking/mappings/ComptrollerLib.ts (hunk 1 above)
Comment (KedziaPawel) [reply subgraphs/diva-staking/mappings/ComptrollerLib.ts]: added, but here I've still used getDepositor in order to be able to track depositors number

File: subgraphs/diva-staking/mappings/ComptrollerLib.ts (hunk 3 above)
Comment (KedziaPawel) [reply subgraphs/diva-staking/mappings/ComptrollerLib.ts]: scaling by shares used right now

File: subgraphs/diva-staking/schema.graphql (hunk 7 above)
Comment (KedziaPawel) [reply subgraphs/diva-staking/schema.graphql]: done

File: subgraphs/diva-staking/utils/vaultHandlers.ts (hunk 9 above)
Comment (KedziaPawel) [reply subgraphs/diva-staking/utils/vaultHandlers.ts]: done

**Comment by KedziaPawel** [comment]
@iherger PR can be re-reviewed

File: subgraphs/diva-staking/mappings/ComptrollerLib.ts (hunk 1 above)
Comment (iherger) [reply subgraphs/diva-staking/mappings/ComptrollerLib.ts]: You can track the depositor numbers also in the creation section of `ensureDepositor`

File: subgraphs/diva-staking/utils/tranches.ts [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  return new Claim(firstClaimAmount, secondClaimAmount);
+}
+
+function getTrancheAccruedRewards(
+  amount: BigDecimal,
+  divPerEthPerDay: BigDecimal,
+  firstStepStakeDays: number,
+  secondStepDaysStaked: number,
```
Comment (iherger) [inline subgraphs/diva-staking/utils/tranches.ts:269]: I would type this as `i32`

Comment (iherger) [reply subgraphs/diva-staking/utils/tranches.ts:269]: Or even `BigDecimal`

**Review by iherger — CHANGES_REQUESTED** [review]

**Review by iherger — APPROVED** [review]

**Pull request merged**